Update guide for GRASS algorithms

AlisterH opened the pull request:
## Description

grass7.txt is woefully out of date - I have attempted to update it, but someone else should check it

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

AlisterH committed:
Update guide for GRASS algorithms

AlisterH commented:
I also note that references to "GRASS 7" are out of date now, at least on Windows...

nicogodet commented on python/plugins/grassprovider/grass7.txt:
```diff
@@ -46,45 +48,57 @@ To indicate the type of geometry, use the following values:
-All outputs are added with the following syntax:
+Raster outputs are added with the following syntax:
+
+QgsProcessingParameterRasterDestination|[name of GRASS output]|[description of output to show]|[Default value, or None]|[True/False, indicating if the parameter is optional or not]
```
Add example ?

AlisterH replied:
added

AlisterH committed:
Update python/plugins/grassprovider/grass7.txt

suggested change from nicogodet

Co-authored-by: Nicolas Godet <39594821+nicogodet@users.noreply.github.com>

AlisterH committed:
Update python/plugins/grassprovider/grass7.txt

suggested change from nicogodet

Co-authored-by: Nicolas Godet <39594821+nicogodet@users.noreply.github.com>

AlisterH committed:
Update grass7.txt

Add more parameters and examples

AlisterH commented on python/plugins/grassprovider/grass7.txt:
```diff
@@ -46,45 +48,57 @@ To indicate the type of geometry, use the following values:
+QgsProcessing.TypeVectorLine
+QgsProcessing.VectorLine
+QgsProcessing.VectorPolygon
+QgsProcessing.TypeVectorAnyGeometry
```
added

AlisterH committed:
Update grass7.txt

Update "advanced" information based on https://medspx.fr/blog/Qgis/grass_processing_provider/

nicogodet commented on python/plugins/grassprovider/grass7.txt:
```diff
@@ -151,22 +151,33 @@ to tag a parameter as "advanced", just add "*" before its declaration. For insta
+
+Example: QgsProcessingParameterFileDestination|html|List of addons|Html files (*.html)|addons_list.html|False
+
+To add additional logic to an algorithm, like a preliminary check on  data, the use of more than one GRASS command,
```
```suggestion
To add additional logic to an algorithm, like a preliminary check on data, the use of more than one GRASS command,
```

nicogodet commented on python/plugins/grassprovider/grass7.txt:
```diff
@@ -151,22 +151,33 @@ to tag a parameter as "advanced", just add "*" before its declaration. For insta
+- Processing the command itself: if you need to chain more than one GRASS command for your algorithm. 
+- Processing the outputs: if you need to do special things before exporting layers or if you need special export methods. 
+
+Whenever you want to add some logic on one (or more) level(s), you have to create a .py file named according to the algorithm name in python/plugins/processing/algs/grass7/ext, replacing '.' with '_'.
```
```suggestion
Whenever you want to add some logic on one (or more) level(s), you have to create a .py file named according to the algorithm name in python/plugins/grassprovider/ext, replacing '.' with '_'.
```

nicogodet commented on python/plugins/grassprovider/grass7.txt:
```diff
@@ -151,22 +151,33 @@ to tag a parameter as "advanced", just add "*" before its declaration. For insta
+- Command: processCommand
+- Outputs: processOutputs
+
+If there is a Python file with the algorithm name in the ext directory, methods will be imported from the file and run instead of the common methods (there are "standard" processCommand/processInputs/processOutputs/checkParameterValuesBeforeExecuting methods in the code of the GRASS provider for QGIS Processing, in python/plugins/processing/algs/grass7/Grass7Algorithm.py).
```
```suggestion
If there is a Python file with the algorithm name in the ext directory, methods will be imported from the file and run instead of the common methods (there are "standard" processCommand/processInputs/processOutputs/checkParameterValuesBeforeExecuting methods in the code of the GRASS provider for QGIS Processing, in python/plugins/grassprovider/Grass7Algorithm.py).
```

nicogodet replied:
```suggestion
If there is a Python file with the algorithm name in the ext directory, methods will be imported from the file and run instead of the common methods (there are "standard" processCommand/processInputs/processOutputs/checkParameterValuesBeforeExecuting methods in the code of the GRASS provider for QGIS Processing, in python/plugins/grassprovider/Grass7Algorithm.py).
```

AlisterH committed:
Update python/plugins/grassprovider/grass7.txt

Suggested change from nicogodet

Co-authored-by: Nicolas Godet <39594821+nicogodet@users.noreply.github.com>

AlisterH committed:
Update python/plugins/grassprovider/grass7.txt

Correction from nicogodet

Co-authored-by: Nicolas Godet <39594821+nicogodet@users.noreply.github.com>

AlisterH committed:
Update python/plugins/grassprovider/grass7.txt

Correction from nicogodet

Co-authored-by: Nicolas Godet <39594821+nicogodet@users.noreply.github.com>

AlisterH committed:
Update grass7.txt

Correct the list of vector output types (guess I was interrupted while working on this section).

AlisterH committed:
Update grass7.txt

add note about reloading descriptions

nyalldawson reviewed (approved):
Great work!
